File: cli/core/session.py
"""
B.DEV CLI Core - Session Management
Persistent state across commands
"""
import json
from pathlib import Path
from datetime import datetime
from typing import Any, Dict, List, Optional

SESSION_DIR = Path.home() / "Dev" / ".bdev" / "cache"
SESSION_FILE = SESSION_DIR / "session.json"
HISTORY_FILE = SESSION_DIR / "history.json"

class Session:
    """Manages session state and command history"""
    
    _instance: Optional['Session'] = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._load()
        return cls._instance
    
    def _load(self):
        SESSION_DIR.mkdir(parents=True, exist_ok=True)
        
        self._state: Dict[str, Any] = {
            "current_project": None,
            "last_command": None,
            "started_at": datetime.now().isoformat(),
            "ai_context": []  # Conversation memory
        }
        
        self._history: List[str] = []
        
        # Load persisted session
        if SESSION_FILE.exists():
            try:
                self._state.update(json.loads(SESSION_FILE.read_text()))
            except:
                pass
        
        # Load history
        if HISTORY_FILE.exists():
            try:
                self._history = json.loads(HISTORY_FILE.read_text())
            except:
                pass
    
    def get(self, key: str, default: Any = None) -> Any:
        return self._state.get(key, default)
    
    def set(self, key: str, value: Any):
        self._state[key] = value
        self._save()
    
    def _save(self):
        SESSION_FILE.write_text(json.dumps(self._state, indent=2, default=str))
    
    # --- Command History ---
    
    def add_to_history(self, command: str):
        self._history.append(command)
        # Keep last 500
        self._history = self._history[-500:]
        HISTORY_FILE.write_text(json.dumps(self._history))
    
    @property
    def history(self) -> List[str]:
        return self._history
    
    # --- AI Conversation Memory ---
    
    def add_ai_message(self, role: str, content: str):
        """Add message to AI conversation memory"""
        self._state.setdefault("ai_context", []).append({
            "role": role,
            "content": content,
            "timestamp": datetime.now().isoformat()
        })
        # Keep last 20 messages
        self._state["ai_context"] = self._state["ai_context"][-20:]
        self._save()
    
    def get_ai_context(self) -> List[Dict]:
        return self._state.get("ai_context", [])
    
    def clear_ai_context(self):
        self._state["ai_context"] = []
        self._save()
    
    # --- Current Project ---
    
    @property
    def current_project(self) -> Optional[Path]:
        p = self._state.get("current_project")
        return Path(p) if p else None
    
    @current_project.setter
    def current_project(self, path: Optional[Path]):
        self._state["current_project"] = str(path) if path else None
        self._save()

# Singleton accessor
def get_session() -> Session:
    return Session()

File: internal/core/session/session.go
package session

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

// Session represents the current CLI session state
type Session struct {
	StartTime    time.Time         `json:"start_time"`
	CurrentDir   string            `json:"current_dir"`
	CommandCount int               `json:"command_count"`
	AIContext    []Message         `json:"ai_context"`
	LastCommand  string            `json:"last_command"`
	Metadata     map[string]string `json:"metadata"`
}

// Message represents an AI conversation message
type Message struct {
	Role      string    `json:"role"`
	Content   string    `json:"content"`
	Timestamp time.Time `json:"timestamp"`
}

// Global session instance
var globalSession *Session

// New creates a new session
func New() *Session {
	cwd, _ := os.Getwd()
	return &Session{
		StartTime:    time.Now(),
		CurrentDir:   cwd,
		CommandCount: 0,
		AIContext:    make([]Message, 0),
		LastCommand:  "",
		Metadata:     make(map[string]string),
	}
}

// Get returns the global session instance
func Get() *Session {
	if globalSession == nil {
		globalSession = New()
	}
	return globalSession
}

// ProjectName returns the current project name from cwd
func (s *Session) ProjectName() string {
	return filepath.Base(s.CurrentDir)
}

// GitBranch returns the current git branch (cached for performance)
func (s *Session) GitBranch() string {
	// Check cache first
	if branch, ok := s.Metadata["git_branch"]; ok {
		// Cache for 5 seconds
		if ts, ok := s.Metadata["git_branch_ts"]; ok {
			if cached, err := time.Parse(time.RFC3339, ts); err == nil {
				if time.Since(cached) < 5*time.Second {
					return branch
				}
			}
		}
	}

	cmd := exec.Command("git", "rev-parse", "--abbrev-ref", "HEAD")
	cmd.Dir = s.CurrentDir
	out, err := cmd.Output()
	if err != nil {
		s.Metadata["git_branch"] = ""
		return ""
	}

	branch := strings.TrimSpace(string(out))
	s.Metadata["git_branch"] = branch
	s.Metadata["git_branch_ts"] = time.Now().Format(time.RFC3339)
	return branch
}

// IsGitRepo checks if current directory is a git repository
func (s *Session) IsGitRepo() bool {
	cmd := exec.Command("git", "rev-parse", "--git-dir")
	cmd.Dir = s.CurrentDir
	return cmd.Run() == nil
}

// IncrementCommandCount increments the command counter
func (s *Session) IncrementCommandCount() {
	s.CommandCount++
}

// SetLastCommand records the last executed command
func (s *Session) SetLastCommand(cmd string) {
	s.LastCommand = cmd
}

// UpdateCurrentDir updates the current working directory
func (s *Session) UpdateCurrentDir() {
	cwd, err := os.Getwd()
	if err == nil {
		s.CurrentDir = cwd
		// Invalidate git cache
		delete(s.Metadata, "git_branch")
		delete(s.Metadata, "git_branch_ts")
	}
}

// AddAIMessage adds a message to AI context
func (s *Session) AddAIMessage(role, content string) {
	s.AIContext = append(s.AIContext, Message{
		Role:      role,
		Content:   content,
		Timestamp: time.Now(),
	})

	// Keep only last 20 messages for context window management
	if len(s.AIContext) > 20 {
		s.AIContext = s.AIContext[len(s.AIContext)-20:]
	}
}

// GetAIContext returns the AI conversation context
func (s *Session) GetAIContext() []Message {
	return s.AIContext
}

// ClearAIContext clears the AI conversation history
func (s *Session) ClearAIContext() {
	s.AIContext = make([]Message, 0)
}

// AIContextSize returns the number of messages in AI context
func (s *Session) AIContextSize() int {
	return len(s.AIContext)
}

// Uptime returns session duration
func (s *Session) Uptime() time.Duration {
	return time.Since(s.StartTime)
}

// UptimeFormatted returns a human-readable uptime
func (s *Session) UptimeFormatted() string {
	d := s.Uptime()

	if d < time.Minute {
		return d.Round(time.Second).String()
	}
	if d < time.Hour {
		return d.Round(time.Minute).String()
	}

	hours := int(d.Hours())
	minutes := int(d.Minutes()) % 60

	if hours > 0 && minutes > 0 {
		return fmt.Sprintf("%dh%dm", hours, minutes)
	} else if hours > 0 {
		return fmt.Sprintf("%dh", hours)
	}
	return fmt.Sprintf("%dm", minutes)
}

// Save persists session to disk
func (s *Session) Save(path string) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}

	data, err := json.MarshalIndent(s, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0o600)
}

// Load restores session from disk
func Load(path string) (*Session, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return New(), nil // File doesn't exist, return new session
	}

	var s Session
	if err := json.Unmarshal(data, &s); err != nil {
		return New(), nil // Invalid JSON, return new session
	}

	// Update current directory
	s.UpdateCurrentDir()

	return &s, nil
}

// Stats returns session statistics
func (s *Session) Stats() map[string]interface{} {
	return map[string]interface{}{
		"uptime":      s.UptimeFormatted(),
		"commands":    s.CommandCount,
		"ai_messages": len(s.AIContext),
		"project":     s.ProjectName(),
		"git_branch":  s.GitBranch(),
		"is_git_repo": s.IsGitRepo(),
	}
}
